### PR TITLE
feat(graph): polished axes, grid, tick labels (#49)

### LINKED ISSUE #49: Polish graph object rendering (axes, labels, grid, curves)
Graph plotter works functionally, but visual output is plain. Needs a styling pass:

- Axis styling: tick marks, readable labels with a subtle halo for on-dark backgrounds.
- Grid: lighter minor lines vs. darker major lines.
- Function curves: antialiased, slight line width for on-screen legibility, distinct stroke styles when multiple expressions plotted.
- Optional: origin emphasis, axis arrows, "nice" axis ranges (snap to 1/2/5 × 10^n).

No behavior change — rendering only. Tracking toward a post-MVP polish PR.

### PR DESCRIPTION
**Pull request by Adamkadaban**
## Summary

Polishes graph object rendering (issue #49) with an adaptive grid, numeric
tick labels, and axis labels. Curves continue to render on top.

## What

- New pure module `src/lib/graph/axes.ts`:
  - `niceStep(range, targetCount)` — picks a 1/2/5×10^k step so the range
    yields ≈ `targetCount` major ticks.
  - `generateTicks(min, max, step?)` — tick values in `[min, max]` at
    multiples of `step` (or a `niceStep` fallback).
  - `formatTick` / `tickDecimals` helpers so labels keep only the precision
    the step actually needs (e.g. `0.25`, not `0.3`).
- `GraphLayer.svelte`:
  - Two-tone grid: light minor lines at `majorStep / 5`, slightly darker
    major lines at the nice step. Falls back to the graph's `gridStep`
    when the user has set one explicitly.
  - Numeric labels on major x/y ticks, "x"/"y" axis labels at the ends of
    the visible axes, and a "0" at the origin when both axes are in view.
  - Curves are drawn after the grid + axes so they sit on top.

## Why

Closes #49. Previous rendering had a single-tone grid tied to a fixed
`gridStep` and no labels, making the graph hard to read across different
zoom levels.

## Testing

- `tests/graph-axes.test.ts` covers `niceStep` for ranges 0.01–1234,
  tick-count bounds, explicit step, degenerate input, and `formatTick` /
  `tickDecimals` behaviour.
- `pnpm lint` ✅
- `pnpm test` ✅ (277 tests)

Closes #49.

### CODE REVIEW

**Comment by Adamkadaban** [comment]
Addressed all five Copilot review comments in 23fdab9:

- **3114409104 (niceStep edge cases, axes.ts:20)** — added `if (!Number.isFinite(rough) || rough <= 0) return 1` after computing `range/targetCount`. Extended `tests/graph-axes.test.ts` with `0`, `-1`, `Infinity`, `NaN`, `Number.MIN_VALUE`, `1e-300 / 1e300` (subnormal underflow), and bad targetCount values.
- **3114409118 (unreachable auto-grid path, GraphLayer.svelte:67)** — `gridStep === 0` now means 'auto'. `createGraphObject` defaults to `0`; `GraphEditor.setGridStep` accepts `0` and still rejects negatives/NaN; `GraphLayer` does `Number.isFinite(g.gridStep) && g.gridStep > 0 ? g.gridStep : niceStep(...)`. Types doc updated.
- **3114409122 (x label clip, GraphLayer.svelte:132)** — `xLabelY` is clamped to `[py + LABEL_PAD, py + ph - LABEL_PAD]` and baseline flips to `top` when the axis is near the top. Tick labels also flip above the axis when it's near the bottom.
- **3114409136 (y label clip, GraphLayer.svelte:145)** — y tick labels clamp their x to `[px + LABEL_PAD, px + pw - LABEL_PAD]` and flip `textAlign` near the left; the 'y' label flips to `right` alignment when the axis is near the right edge.
- **3114409148 (grid explosion, GraphLayer.svelte:83)** — new `cappedStep(range, step, maxLines)` helper in `axes.ts` (unit tested). Rendering uses `MAX_GRID_LINES_PER_AXIS = 200`: major step is coarsened through `cappedStep`, and minor lines are skipped when `range / minorStep > 200`.

`pnpm lint` and `pnpm test` (281 tests) both pass.

File: src/lib/graph/axes.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ */
+export function niceStep(range: number, targetCount: number): number {
+  if (!Number.isFinite(range) || range <= 0) return 1;
+  if (!Number.isFinite(targetCount) || targetCount <= 0) return range;
+
+  const rough = range / targetCount;
+  const exponent = Math.floor(Math.log10(rough));
+  const pow = 10 ** exponent;
+  const mantissa = rough / pow;
+
```
[the remark this answers is not in the document]
Comment (Adamkadaban) [reply src/lib/graph/axes.ts:21]: Fixed. Guarded `rough = range / targetCount` for non-finite / non-positive; returns `1` as a safe default. Added test coverage for range=0, negative, `Number.MIN_VALUE`, `Infinity`, `NaN`, and degenerate targetCount. Commit 23fdab9.

File: src/lib/canvas/GraphLayer.svelte [hunk 2]
```diff
@@ -54,40 +61,95 @@
     const xToPx = (x: number) => px + ((x - x0) / xSpan) * pw;
     const yToPx = (y: number) => py + (1 - (y - y0) / ySpan) * ph;
 
-    if (g.showGrid && g.gridStep > 0) {
-      ctx.strokeStyle = GRID_COLOR;
+    const majorXStep = g.gridStep > 0 ? g.gridStep : niceStep(xSpan, TARGET_MAJOR_TICKS);
+    const majorYStep = g.gridStep > 0 ? g.gridStep : niceStep(ySpan, TARGET_MAJOR_TICKS);
+    const minorXStep = majorXStep / MINOR_SUBDIVISIONS;
+    const minorYStep = majorYStep / MINOR_SUBDIVISIONS;
```
[the remark this answers is not in the document]
Comment (Adamkadaban) [reply src/lib/canvas/GraphLayer.svelte]: Fixed via a `gridStep === 0` sentinel for auto mode. `createGraphObject` defaults to 0, `GraphEditor.setGridStep` now accepts 0 (rejects negatives/NaN), and the renderer falls back to `niceStep` when `gridStep <= 0` or non-finite. Commit 23fdab9.

File: src/lib/canvas/GraphLayer.svelte [hunk 3]
```diff
@@ -54,40 +61,95 @@
...
+        const axisY = yToPx(0);
+        ctx.textAlign = 'center';
+        ctx.textBaseline = 'top';
+        for (const x of generateTicks(x0, x1, majorXStep)) {
+          if (yAxisVisible && Math.abs(x) < majorXStep * 1e-6) continue;
+          ctx.fillText(formatTick(x, majorXStep), xToPx(x), axisY + LABEL_PAD);
+        }
+        ctx.textAlign = 'right';
+        ctx.textBaseline = 'alphabetic';
+        ctx.fillText('x', px + pw - LABEL_PAD, axisY - LABEL_PAD);
```
[the remark this answers is not in the document]
Comment (Adamkadaban) [reply src/lib/canvas/GraphLayer.svelte]: Fixed. `xLabelY` is now clamped inside `[py + LABEL_PAD, py + ph - LABEL_PAD]` and baseline flips to `top` when the axis is near the top edge (same for tick labels). Commit 23fdab9.

File: src/lib/canvas/GraphLayer.svelte [hunk 4]
```diff
@@ -54,40 +61,95 @@
...
+        const axisX = xToPx(0);
+        ctx.textAlign = 'right';
+        ctx.textBaseline = 'middle';
+        for (const y of generateTicks(y0, y1, majorYStep)) {
+          if (xAxisVisible && Math.abs(y) < majorYStep * 1e-6) continue;
+          ctx.fillText(formatTick(y, majorYStep), axisX - LABEL_PAD, yToPx(y));
+        }
+        ctx.textAlign = 'left';
+        ctx.textBaseline = 'top';
+        ctx.fillText('y', axisX + LABEL_PAD, py + LABEL_PAD);
```
[the remark this answers is not in the document]
Comment (Adamkadaban) [reply src/lib/canvas/GraphLayer.svelte]: Fixed. Y-axis tick label x-coordinates are now clamped inside the frame and `textAlign` flips between `left`/`right` based on proximity to the left/right edges. 'y' axis label flips alignment near the right edge. Commit 23fdab9.

File: src/lib/canvas/GraphLayer.svelte [hunk 5]
```diff
@@ -54,40 +61,95 @@
...
+        const sx = xToPx(x);
+        ctx.moveTo(sx, py);
+        ctx.lineTo(sx, py + ph);
+      }
+      for (const y of generateTicks(y0, y1, minorYStep)) {
+        const sy = yToPx(y);
+        ctx.moveTo(px, sy);
+        ctx.lineTo(px + pw, sy);
+      }
+      ctx.stroke();
```
[the remark this answers is not in the document]
Comment (Adamkadaban) [reply src/lib/canvas/GraphLayer.svelte]: Fixed. Added `cappedStep(range, step, maxLines)` helper in `axes.ts` (unit-tested) and applied `MAX_GRID_LINES_PER_AXIS = 200`. Minor lines are skipped per axis when `range / minorStep > 200`, major lines coarsen to stay under the cap. Prevents hangs with tiny user grid steps. Commit 23fdab9.
